### PR TITLE
Mat 8438 aggregate measure history data

### PR DESCRIPTION
**Pull request by sb-prateekkeerthi**
## MADiE PR

Jira Ticket: [MAT-8438](https://jira.cms.gov/browse/MAT-8438)
(Optional) Related Tickets:

### Summary
aggregate measure history data

### All Submissions
* [ ] This PR has the JIRA linked.
* [ ] Required tests are included.
* [ ] No extemporaneous files are included (i.e Complied files or testing results).
* [ ] This PR is merging into the **correct branch**.
* [ ] All Documentation needed for this PR is Complete (or noted in a TODO or other Ticket).
* [ ] Any breaking changes or failing automations are noted by placing a comment on this PR.

### DevSecOps
If there is a question if this PR has a security or infrastructure impact, please contact the Security or DevOps engineer assigned to this project to discuss it further.

* [ ] This PR has NO significant security impact (i.e Changing auth methods, Adding a new user type, Adding a required but vulnerable package).

### Reviewers
By Approving this PR you are attesting to the following:

*  Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose.
*  The tests appropriately test the new code, including edge cases.
*  If you have any concerns they are brought up either to the developer assigned, security engineer, or leads.

### CODE REVIEW

File: src/main/java/cms/gov/madie/measure/resources/MeasureController.java [hunk 1]
```diff
@@ -454,4 +455,11 @@ public ResponseEntity<Boolean> transferMeasures(
         measureService.transferMeasures(
             measureIds, harpId, retainShareAccess, principal.getName()));
   }
+
+  @GetMapping(value = "/measures/measure-history/{id}")
```
Comment (jkotanchik-SB) [inline src/main/java/cms/gov/madie/measure/resources/MeasureController.java]: How about this url?
```suggestion
  @GetMapping(value = "/measures/{id}/history")
```

File: src/main/java/cms/gov/madie/measure/services/ActionLogService.java [hunk 2]
```diff
@@ -84,4 +85,30 @@ public boolean logMeasureSetAction(
             .build(),
         collection);
   }
+
+  public List<Action> findMeasureHistory(String targetId, String measureSetId) {
```
Comment (jkotanchik-SB) [inline src/main/java/cms/gov/madie/measure/services/ActionLogService.java]:
```suggestion
  public List<Action> findMeasureHistory(String measureId, String measureSetId) {
```
Recommend naming the parameter `measureId` to help the caller better understand what's needed.

File: src/main/java/cms/gov/madie/measure/repositories/MeasureActionLogRepository.java [hunk 3]
```diff
@@ -3,5 +3,9 @@
 import gov.cms.madie.models.common.ActionLog;
 import org.springframework.data.mongodb.repository.MongoRepository;
 
+import java.util.List;
+
 public interface MeasureActionLogRepository
-    extends MongoRepository<ActionLog, String>, ActionLogRepository {}
+    extends MongoRepository<ActionLog, String>, ActionLogRepository {
+  List<ActionLog> findByTargetId(String targetId);
```
Comment (jkotanchik-SB) [inline src/main/java/cms/gov/madie/measure/repositories/MeasureActionLogRepository.java]: Are we expecting a Measure instance to match multiple Log documents?

Related, how about a Measure Set?

**Review by jkotanchik-SB — CHANGES_REQUESTED** [review]
Update method parameter, and maybe the url.

File: src/main/java/cms/gov/madie/measure/resources/MeasureController.java (hunk 1 above)
Comment (sb-prateekkeerthi) [reply src/main/java/cms/gov/madie/measure/resources/MeasureController.java]: Sure, i'll make the change

File: src/main/java/cms/gov/madie/measure/services/ActionLogService.java (hunk 2 above)
Comment (sb-prateekkeerthi) [reply src/main/java/cms/gov/madie/measure/services/ActionLogService.java]: Sure, i'll make the change

File: src/main/java/cms/gov/madie/measure/repositories/MeasureActionLogRepository.java (hunk 3 above)
Comment (sb-prateekkeerthi) [reply src/main/java/cms/gov/madie/measure/repositories/MeasureActionLogRepository.java]: fixed it, thank you

File: src/main/java/cms/gov/madie/measure/services/ActionLogService.java [hunk 4]
```diff
@@ -85,4 +86,25 @@ public boolean logMeasureSetAction(
...
+
+    List<Action> combinedActionLogs = new ArrayList<>();
+
+    if (measureActionLogs != null && !CollectionUtils.isEmpty(measureActionLogs.getActions())) {
+      combinedActionLogs.addAll(measureActionLogs.getActions());
+    }
+
+    measureSetActionLogs.ifPresent(
+        log -> {
+          if (log.getActions() != null && !log.getActions().isEmpty()) {
```
Comment (jkotanchik-SB) [inline src/main/java/cms/gov/madie/measure/services/ActionLogService.java]:
```suggestion
          if (CollectionUtils.isNotEmpty(log.getActions())) {
```
Recommend simplifying with CollectionUtils.

File: src/main/java/cms/gov/madie/measure/services/ActionLogService.java [hunk 5]
```diff
@@ -85,4 +86,25 @@ public boolean logMeasureSetAction(
             .build(),
         collection);
   }
+
+  public List<Action> findMeasureHistory(String measureId, String measureSetId) {
+    ActionLog measureActionLogs = measureActionLogRepository.findByTargetId(measureId);
+    Optional<MeasureSetActionLog> measureSetActionLogs =
```
Comment (jkotanchik-SB) [inline src/main/java/cms/gov/madie/measure/services/ActionLogService.java]: It's a little odd that only one of the repos returns an `Optional`. Would it be problematic to have `measureActionLogRepository.findByTargetId` also return an `Optional`?

File: src/main/java/cms/gov/madie/measure/services/ActionLogService.java (hunk 4 above)
Comment (sb-prateekkeerthi) [reply src/main/java/cms/gov/madie/measure/services/ActionLogService.java]: done

File: src/main/java/cms/gov/madie/measure/services/ActionLogService.java (hunk 5 above)
Comment (sb-prateekkeerthi) [reply src/main/java/cms/gov/madie/measure/services/ActionLogService.java]: I had the same question but this was something which was already there so I didn't make any change at that time, but yeah now I have fixed as you also suggested that

**Review by jkotanchik-SB — APPROVED** [review]
